### ISSUE TITLE
Constrained Optimization as a Regularizer

### ISSUE THREAD

**Issue by cancan101** [benanne/Lasagne#84]
Consider adding constrained optimization as a regularizer.

From: https://blog.safaribooksonline.com/2014/02/13/constrained-optimization-pylearn2/ one of the regularizers that pyearn2 supports is `col_norm`.

See also: http://deeplearning.net/software/pylearn2/library/models.html?highlight=linear#pylearn2.models.mlp.Linear.

**Comment by benanne** [comment]
Good idea. I've never had much luck with this approach in the past, but I might end up implementing it myself anyway, eventually. Pull requests are welcome, at any rate :)

**Comment by ebattenberg** [comment]
I've used weight norm constraints in the past.  I think they can be useful if you have some filters (or rows/cols) in the lower layers that tend to die with weight decay.  

I have some Theano code that does this, and I'd be willing to add it in a PR.  

Since just setting a maximum value for the norm by hand can be a bit tricky, I tend to use a multiple of the initial norm of the layer as the argument that ends up setting the constraint.  For example, if the incoming weights are initialized with standard deviation `std`, then the expected incoming 2-norm of an incoming weight vector is `std*sqrt(fan_in)`.  The argument I use is a multiple of this value (something between 1x and ~5x) depending on the nonlinearity used and how close to the output layer we are.  

Of course then the weight norm constraint would need information about the initialization values, but I guess you could pass the initialization type information through.  

Does that sound a reasonable way to hyperparametrize this?  I can also allow just the exact constraint as well, but I don't know if that is as meaningful to the user.  I hate having to recompute the constraints when layer sizes change.

**Comment by f0k** [comment]
> Does that sound like a reasonable way to hyperparametrize this?

Sounds interesting.

> Of course then the weight norm constraint would need information about the initialization values, but I guess you could pass the initialization type information through.

If the regularization is applied directly to shared variables rather than more complex expressions, it could just check the values of said variables (e.g., their initial column norms) on construction. For more complex expressions, that would still be possible via an `eval()` call. Seems easier and more flexible to me than passing on the weight initializer.

> I can also allow just the exact constraint as well

It should probably allow both, maybe as `max_col_norm=somevalue` and `max_col_norm='auto'`.

We should probably rework the regularization code first, though (see #14).

**Comment by ebattenberg** [comment]
> It should probably allow both, maybe as `max_col_norm=somevalue` and `max_col_norm='auto'.`

The way I had in mind for doing the version that is a multiple of the initial average norm is by providing a parameter for the multiple.  So, it wouldn't be purely automatic.  Just the initial average norm would be computed automatically and the constraint would be set as a multiple of this given the provided multiplier value.  

Any suggestions on the best way to allow people to configure either the absolute constraint or the multiplier/relative constraint?  

```
def weight_norm_constraint(layer, constraint_type='multiplier', constraint_value=1.0)
```
Then there could be an `'absolute` `constraint_type`, where the passed `constrain_value` is the actual constraint.  That doesn't seem very nice.  

There could also just be two different kwargs:
```
def weight_norm_constraint(layer, absolute_constraint=None, multiplier_constraint=1.0):        
```
And then the `absolute_constraint` is used if provided, otherwise we use the `multiplier_constrain`.


> We should probably rework the regularization code first, though (see #14).

Agreed.  Just starting the brainstorming, and making sure we are considering different types of regularizers.

**Comment by f0k** [comment]
> `def weight_norm_constraint(layer`

This should be `norm_constraint(param`, so one can finely control what this is applied to: Particular weights of particular layers, or anything else that is a shared variable.

> There could also just be two different kwargs: [...]
> And then the `absolute_constraint` is used if provided, otherwise we use the `multiplier_constrain`.

What about two kwargs and using both, similar to how `np.allclose()` has a `rtol` and `atol` for relative and absolute difference that are used both? E.g.:
```python
def norm_constraint(param, abs_max=None, rel_max=None, axis=0)
```
which would constrain the column norms to `abs_max` or `rel_max * initial norm`, whichever is stricter.

> Agreed. Just starting the brainstorming, and making sure we are considering different types of regularizers.

You're right, I forgot that we're talking about constrained optimization that needs to directly affect the `updates` dictionary. This means we may need a different signature for that, depending on how we want to use it later. With the signature I proposed above it could be used like:
```python
updates = lasagne.updates.nesterov(loss, params)
updates[someparam] = norm_constraint(someparam, abs_max=15)
```
Are there any constrained optimization methods that need to maintain state, i.e., that need to create additional shared variables with update rules? In this case, the signature should be:
```python
updates = lasagne.updates.nesterov(loss, params)
updates = norm_constraint(updates, someparam, abs_max=15)
```
By the way, if we call it `norm_constraint`, it should probably also allow constraining the norm from below, or constraining it to a fixed value (as done for the factor-to-hidden matrix in mcRBMs, for example). That would just need an additional `abs_min` and `rel_min` parameter.

**Comment by benanne** [comment]
> ```
updates = lasagne.updates.nesterov(loss, params)
updates[someparam] = norm_constraint(someparam, abs_max=15)
```

That would probably have to be:

```
updates = lasagne.updates.nesterov(loss, params)
updates[someparam] = norm_constraint(updates[someparam], abs_max=15)
```

Else the parameter doesn't actually get updated, only constrained :)
This does look a bit cumbersome, maybe we can provide a function that applies it to all elements in the dictionary as a shortcut?

**Comment by f0k** [comment]
> Else the parameter doesn't actually get updated, only constrained :)

Ooops, yes. It's late. So we need the other syntax:
```python
updates = lasagne.updates.nesterov(loss, params)
updates = norm_constraint(updates, someparam, abs_max=15)
```
Not only because it will prevent errors by tired users, but also because Eric needs `someparam` and not just `updates[someparam]` to be able to determine the shared variable's original column norm. (Alternatively, we could have the signature take `updates[someparam]` to censor it and treat `norm_constraint()` as a special case that also needs `someparam` itself in case `rel_max` is given.)

> This does look a bit cumbersome, maybe we can provide a function that applies it to all elements in the dictionary as a shortcut?

No, you probably don't want to apply it to the bias vectors and the momentum/velocity terms that are part of the dictionary. There could be a shortcut that applies it to all weights in a network (given the output layer(s)), though, similar to what I proposed for regularization in https://github.com/benanne/Lasagne/issues/14#issuecomment-55393028.


We will need to very clearly distinguish between regularization costs (that take an expression and return an expression that can be added to the cost before differentiating it) and constraints / censored updates (that take an updates dictionary and return an altered updates dictionary). We could place both in `lasagne.regularization`, or we could put the former in `lasagne.objectives` and the latter in `lasagne.updates`, or we could think of something else. It's a good time to discuss this now, before any of this has actually been added to Lasagne.

**Comment by benanne** [comment]
> No, you probably don't want to apply it to the bias vectors and the momentum/velocity terms that are part of the dictionary. There could be a shortcut that applies it to all weights in a network (given the output layer(s)), though, similar to what I proposed for regularization in #14 (comment).

That makes sense, I made the assumption that the user would be responsible for making that distinction (i.e. they would have to make two separate update dictionaries, one for the weight params and one for the bias params, then apply the norm constraint only to the weight params, and then merge them).

But a convenience function for this is definitely a good idea.

> We will need to very clearly distinguish between regularization costs (that take an expression and return an expression that can be added to the cost before differentiating it) and constraints / censored updates (that take an updates dictionary and return an altered updates dictionary). We could place both in `lasagne.regularization`, or we could put the former in `lasagne.objectives` and the latter in `lasagne.updates`, or we could think of something else. It's a good time to discuss this now, before any of this has actually been added to Lasagne.

Agreed. There is also a third class of regularization strategies, those that inject noise (like dropout). We already have a `DropoutLayer` in a `layers` submodule (`noise`), so that's an argument for adding both of these types of things under `objectives` and `updates` respectively. We can always divide up those modules further when appropriate anyway.

**Comment by f0k** [comment]
> I made the assumption that the user would be responsible for making that distinction (i.e. they would have to make two separate update dictionaries, one for the weight params and one for the bias params, then apply the norm constraint only to the weight params, and then merge them).

Depending on the learning scheme, half of the weight param updates dictionary will still need to be excluded from the constraints (e.g., the velocity terms). (It probably wouldn't affect learning for `abs_max=15`, just hamper performance, but it would be very bad for `rel_max=5` since the velocity terms are initialized to all zeros.)

> Agreed. There is also a third class of regularization strategies, those that inject noise (like dropout).

Ah, good point. I'm strongly in favor of adding regularization costs to `objectives` and update constraints to `updates` or `updates.constraints` then (since update constraints have a different interface to existing updates).

**Comment by ebattenberg** [comment]
> I'm strongly in favor of adding regularization costs to objectives and update constraints to updates or updates.constraints then (since update constraints have a different interface to existing updates).

I can see the reasons for having the different types of regularization in different submodules based on their underlying implementation.  This makes sense from a developer standpoint.  But from a user standpoint, would it makes things a bit more confusing to have to look for different regularizers in different locations?  Personally, I think of weight decay and weight norm constraints as two related alternatives.
